daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/ledgerinteraction/JsonLedgerClient.scala

```diff
@@ -294,29 +313,41 @@ class JsonLedgerClient(
     )
   }
 
-  // Check that the parties in the token match the given parties.
-  private def validateTokenParties(parties: OneAnd[Set, Ref.Party], what: String): Future[Unit] = {
-    import scalaz.std.string._
-    val tokenParties = Set(tokenPayload.readAs ++ tokenPayload.actAs: _*)
+  // Check that the parties in the token provide read claims for the given parties
+  // and return explicit party specifications if required.
+  private def validateTokenParties(
+      parties: OneAnd[Set, Ref.Party],
+      what: String,
+  ): Future[Option[QueryParties]] = {
+    val tokenParties = tokenPayload.readAs.toSet union tokenPayload.actAs.toSet
+    // Use toSet twice to convert from Party to String
```

Comment:
```suggestion
```
Either everywhere, or nowhere. We probably does not want to add this comment everywhere.